File: ios/tv_textfield/Sources/tv_textfield/TvTextFieldPlatformView.swift
import Flutter
import UIKit

class TvTextFieldViewFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    return TvTextFieldPlatformView(
      frame: frame,
      viewId: viewId,
      args: args as? [String: Any],
      messenger: messenger
    )
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}

class TvTextFieldPlatformView: NSObject, FlutterPlatformView, UITextFieldDelegate {
  private let textField = UITextField(frame: .zero)
  private let channel: FlutterMethodChannel
  private var params: [String: Any]

  init(
    frame: CGRect,
    viewId: Int64,
    args: [String: Any]?,
    messenger: FlutterBinaryMessenger
  ) {
    channel = FlutterMethodChannel(
      name: "tv_textfield/edit_text_\(viewId)",
      binaryMessenger: messenger
    )
    params = args ?? [:]
    super.init()
    textField.frame = frame
    textField.delegate = self
    textField.borderStyle = .none
    textField.addTarget(self, action: #selector(textDidChange), for: .editingChanged)
    applyParams()
    channel.setMethodCallHandler(handle)
  }

  func view() -> UIView {
    return textField
  }

  private func applyParams() {
    textField.text = params["text"] as? String ?? ""
    textField.placeholder = params["hint"] as? String
    textField.isEnabled = params["enabled"] as? Bool ?? true
    textField.isSecureTextEntry = params["obscureText"] as? Bool ?? false
    textField.textAlignment = mapTextAlign(params["textAlign"] as? String)
    if let colorValue = params["textColor"] as? Int {
      textField.textColor = UIColor(argb: colorValue)
    } else {
      textField.textColor = .label
    }
    textField.returnKeyType = mapReturnKey(params["textInputAction"] as? String)
    textField.keyboardType = mapKeyboardType(params["keyboardType"] as? String)
    if params["autofocus"] as? Bool == true {
      textField.becomeFirstResponder()
    }
  }

  private func mapTextAlign(_ name: String?) -> NSTextAlignment {
    switch name {
    case "center": return .center
    case "right", "end": return .right
    default: return .left
    }
  }

  private func mapReturnKey(_ name: String?) -> UIReturnKeyType {
    switch name {
    case "done": return .done
    case "go": return .go
    case "next": return .next
    case "search": return .search
    case "send": return .send
    default: return .default
    }
  }

  private func mapKeyboardType(_ name: String?) -> UIKeyboardType {
    switch name {
    case "number": return .numberPad
    case "phone": return .phonePad
    case "emailAddress": return .emailAddress
    case "url": return .URL
    case "visiblePassword": return .asciiCapable
    case "multiline": return .default
    default: return .default
    }
  }

  @objc private func textDidChange() {
    channel.invokeMethod("onTextChanged", arguments: textField.text ?? "")
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setText":
      let text = call.arguments as? String ?? ""
      if textField.text != text {
        textField.text = text
      }
      result(nil)
    case "setFocused":
      let focused = call.arguments as? Bool ?? false
      if focused {
        textField.becomeFirstResponder()
      } else {
        textField.resignFirstResponder()
      }
      result(nil)
    case "update":
      if let updated = call.arguments as? [String: Any] {
        params = updated
        applyParams()
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func textFieldDidBeginEditing(_ textField: UITextField) {
    channel.invokeMethod("onFocusChanged", arguments: true)
  }

  func textFieldDidEndEditing(_ textField: UITextField) {
    channel.invokeMethod("onFocusChanged", arguments: false)
  }

  func textFieldShouldReturn(_ textField: UITextField) -> Bool {
    channel.invokeMethod("onSubmitted", arguments: textField.text ?? "")
    textField.resignFirstResponder()
    return true
  }
}

private extension UIColor {
  convenience init(argb: Int) {
    let alpha = CGFloat((argb >> 24) & 0xFF) / 255.0
    let red = CGFloat((argb >> 16) & 0xFF) / 255.0
    let green = CGFloat((argb >> 8) & 0xFF) / 255.0
    let blue = CGFloat(argb & 0xFF) / 255.0
    self.init(red: red, green: green, blue: blue, alpha: alpha)
  }
}

File: android/src/main/kotlin/com/tvtextfield/tv_textfield/TvTextFieldPlatformView.kt
package com.tvtextfield.tv_textfield

import android.graphics.Color
import android.graphics.Typeface
import android.text.InputType
import android.view.Gravity
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class TvTextFieldFactory(private val messenger: BinaryMessenger) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: android.content.Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val params = args as? Map<String, Any?> ?: emptyMap()
        return TvTextFieldPlatformView(context, messenger, viewId, params)
    }
}

class TvTextFieldPlatformView(
    context: android.content.Context,
    messenger: BinaryMessenger,
    viewId: Int,
    private var params: Map<String, Any?>,
) : PlatformView, MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, "tv_textfield/edit_text_$viewId")
    private val editText = EditText(context)

    init {
        channel.setMethodCallHandler(this)
        applyParams()
        editText.setOnFocusChangeListener { _, hasFocus ->
            channel.invokeMethod("onFocusChanged", hasFocus)
        }
        editText.addTextChangedListener(object : android.text.TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: android.text.Editable?) {
                if (suppressTextEvents) {
                    return
                }
                channel.invokeMethod("onTextChanged", s?.toString() ?: "")
            }
        })
        editText.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE ||
                actionId == EditorInfo.IME_ACTION_GO ||
                actionId == EditorInfo.IME_ACTION_NEXT ||
                actionId == EditorInfo.IME_ACTION_SEARCH ||
                actionId == EditorInfo.IME_ACTION_SEND
            ) {
                channel.invokeMethod("onSubmitted", editText.text?.toString() ?: "")
                true
            } else {
                false
            }
        }
        editText.isFocusable = true
        editText.isFocusableInTouchMode = true
    }

    private var suppressTextEvents = false

    private fun applyParams() {
        editText.hint = params["hint"] as? String ?: ""
        val text = params["text"] as? String ?: ""
        if (editText.text.toString() != text) {
            suppressTextEvents = true
            try {
                editText.setText(text)
                editText.setSelection(text.length)
            } finally {
                suppressTextEvents = false
            }
        }
        editText.isEnabled = params["enabled"] as? Boolean ?: true
        editText.inputType = mapKeyboardType(params["keyboardType"] as? String)
        if (params["obscureText"] as? Boolean == true) {
            editText.inputType =
                editText.inputType or InputType.TYPE_TEXT_VARIATION_PASSWORD
            editText.transformationMethod =
                android.text.method.PasswordTransformationMethod.getInstance()
        }
        val textColor = params["textColor"] as? Number
        if (textColor != null) {
            editText.setTextColor(textColor.toInt())
        } else {
            editText.setTextColor(Color.WHITE)
        }
        editText.setHintTextColor(Color.GRAY)
        editText.setBackgroundColor(Color.TRANSPARENT)
        editText.typeface = Typeface.DEFAULT
        editText.maxLines = (params["maxLines"] as? Number)?.toInt() ?: 1
        editText.minLines = (params["minLines"] as? Number)?.toInt() ?: 1
        editText.gravity = mapTextAlign(params["textAlign"] as? String)
        if (params["autofocus"] as? Boolean == true) {
            editText.requestFocus()
        }
        mapImeAction(params["textInputAction"] as? String)?.let { editText.imeOptions = it }
    }

    private fun mapKeyboardType(name: String?): Int {
        return when (name) {
            "number" -> InputType.TYPE_CLASS_NUMBER
            "phone" -> InputType.TYPE_CLASS_PHONE
            "emailAddress" -> InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS
            "url" -> InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_URI
            "visiblePassword" -> InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
            "multiline" -> InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_MULTI_LINE
            else -> InputType.TYPE_CLASS_TEXT
        }
    }

    private fun mapTextAlign(name: String?): Int {
        return when (name) {
            "center" -> Gravity.CENTER
            "right", "end" -> Gravity.CENTER_VERTICAL or Gravity.END
            else -> Gravity.CENTER_VERTICAL or Gravity.START
        }
    }

    private fun mapImeAction(name: String?): Int? {
        return when (name) {
            "done" -> EditorInfo.IME_ACTION_DONE
            "go" -> EditorInfo.IME_ACTION_GO
            "next" -> EditorInfo.IME_ACTION_NEXT
            "search" -> EditorInfo.IME_ACTION_SEARCH
            "send" -> EditorInfo.IME_ACTION_SEND
            else -> null
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setText" -> {
                val text = call.arguments as? String ?: ""
                if (editText.text.toString() != text) {
                    suppressTextEvents = true
                    try {
                        editText.setText(text)
                        editText.setSelection(text.length)
                    } finally {
                        suppressTextEvents = false
                    }
                }
                result.success(null)
            }
            "setFocused" -> {
                val focused = call.arguments as? Boolean ?: false
                if (focused) {
                    editText.requestFocus()
                } else {
                    editText.clearFocus()
                }
                result.success(null)
            }
            "update" -> {
                @Suppress("UNCHECKED_CAST")
                params = call.arguments as? Map<String, Any?> ?: params
                applyParams()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun getView(): View = editText

    override fun dispose() {
        channel.setMethodCallHandler(null)
    }
}
